### ISSUE TITLE
false positive in rule "newline-after-var"

### ISSUE THREAD

**Issue by doberkofler** [eslint/eslint#2237]
When using eslint v0.18.0 on the windows platform the following code example:
```javascript
/*eslint-env node*/
/*eslint newline-after-var:[2, "always"]*/

"use strict";

var test = require("selenium-webdriver/testing");

test.describe("Widget - User defined", function () {
	var client;

	//	Prepare the tests

	this.timeout(client);
});
```

shows the following false positives:
```
6:0  error  Newline is always expected after a "var" statement  newline-after-var
9:1  error  Newline is always expected after a "var" statement  newline-after-var
```

**Comment by cvisco** [comment]
This should be fixed by pull request https://github.com/eslint/eslint/pull/2236, which was merged today. It completely re-implements how the `newline-after-var` rule works to prevent it from issuing false positives on excess whitespace or comments after the var declaration.

**Comment by doberkofler** [comment]
Excellent! Did I understand correctly, that the following example would therefore also no longer issue an error:
```javascript
/*eslint-env node*/
/*eslint newline-after-var:[2, "always"]*/

"use strict";

var test = require("selenium-webdriver/testing");

test.describe("Widget - User defined", function () {
    var client;
    //  Prepare the tests
    this.timeout(client);
});
```

**Comment by cvisco** [comment]
@doberkofler: Yes, under the newer implementation, the rule is looking at where the next actual token lands. Given that `var client;` in your example is on line 9 and the next actual token is `this` on line 11, it would pass. Now that I look at it, though, i'm not sure that's the right way to go. Personally, I would expect this to be a warning since there is no actual blank line after the `var`. What is your expectation in this case?

The previous implementation used a RegEx pattern to determine if a newline immediately followed the `var` declaration. Unfortunately that was causing false positives on extra whitespace at the end of the line (which was the likely cause of your original issue) and also inline comments at the end of the line, which is what issue #2229 was about.

Given that the new implementation is comment-tolerant, I may have introduced a slight regression in the particular use case you brought up. If this gets accepted as a bug, I'll add tests around this pattern and patch the rule to make sure the following line is, in fact, empty (whitespace characters only).

**Comment by nzakas** [comment]
So what is the problem being reported with this issue?

**Comment by cvisco** [comment]
@nzakas: The actual problem appears in @doberkofler's second comment above, and that would be with the pattern:
```js
/*eslint newline-after-var:[2, "always"]*/

var foo = "bar";
//  This will not cause a warning
doSomethingWith(foo)
```
My recent update to `newline-after-var` will not warn in this case because it looks at the placement of tokens and doesn't really see the comment. This is a regression that I introduced in c1ceb6050bae9245c4562fe22df35c30f674dddf. I can fix this tonight.

**Comment by doberkofler** [comment]
1) I tested your commit and it fixed my original problem where false positive was reported although there was a newline.

2) The second questions seems to be a regression compared to the documentation and the original implementation. Based on my experience with this rule, I would recommend to also allow the new behavior by adding an option that allows to use a comment instead of a newline.

**Comment by nzakas** [comment]
@cvisco thanks for the explanation

**Comment by cvisco** [comment]
@doberkofler: I'd like to focus on the regression bug here, as that'll be important to fix before the next patch release. As for optionally allowing comments in lieu of blank lines, I would guess it might make sense to file a separate enhancement issue for that.

**Comment by nzakas** [comment]
Yes, one request per issue, please. Makes it a lot easier to track things. :)

**Comment by cvisco** [comment]
I developed a fix for the regression last night. Just want to expand the test coverage tonight to catch a lot more potential corner cases like this, and then I will submit a PR
